R/template.R

```diff
@@ -0,0 +1,123 @@
+#' Process an HTML template
+#'
+#' Process an HTML template and return a tagList object. If the template is a
+#' complete HTML document, then the returned object will also have class
+#' \code{html_document}, and can be passed to the function
+#' \code{\link{renderDocument}} to get the final HTML text.
+#'
+#' @param filename Path to an HTML template file.
+#' @param ... Variable values to use when processing the template.
+#' @param document_ Is this template a complete HTML document (\code{TRUE}), or
+#'   a fragment of HTML that is to be inserted into an HTML document
+#'   (\code{FALSE})? With \code{"auto"} (the default), auto-detect by searching
+#'   for the string \code{"<HTML>"} within the template.
+#'
+#' @seealso \code{\link{renderDocument}}
+#' @export
+htmlTemplate <- function(filename, ..., document_ = "auto") {
+  html <- readChar(filename, file.info(filename)$size, useBytes = TRUE)
+
+  pieces <- strsplit(html, "{{", fixed = TRUE)[[1]]
+  pieces <- strsplit(pieces, "}}", fixed = TRUE)
+
+  # Each item in `pieces` is a 2-element character vector. In that vector, the
+  # first item is code, and the second is text. The one exception is that the
+  # first item in `pieces` will be a 1-element char vector; that element is
+  # text.
+  if (length(pieces[[1]]) != 1) {
+    stop("Mismatched {{ and }} in HTML template.")
+  }
+  lapply(pieces[-1], function(x) {
+    if (length(x) != 2) {
+      stop("Mismatched {{ and }} in HTML template.")
+    }
+  })
+
+  # Create environment to evaluate code, as a child of the global env. This
+  # environment gets the ... arguments assigned as variables.
+  vars <- list(...)
+  if ("headContent" %in% names(vars)) {
+    stop("Can't use reserved argument name 'headContent'.")
+  }
+  vars$headContent <- function() HTML("<!-- HEAD_CONTENT -->")
+  env <- list2env(vars, parent = globalenv())
```

Comment:
Would that mean that local variables in the caller will become available to the template (i.e. dynamic scoping)? If so, we don't want that. We only want access to 1) the explicitly passed in vars (`...`), 2) the global environment, and 3) the search path.